lib/logical_solver/presolve.cpp

```diff
@@ -0,0 +1,889 @@
+      if (presolve_active_rows_.at(i)) {
+
+        for (int j = 0; j < variables_count_; ++j) {
+          if (presolve_active_columns_.at(j)) {
+
+            if (problem_matrix_.at(i).at(j) != 0) {
+              rows_non_zero_variables_.at(i).push_back(j);
+              cols_non_zeros_indices_.at(j).push_back(i);
+            }
+          }
+        }
+      }
+    }
+  };
+
+  double Presolve::getFeasibleValueCalculationBound(int row_index) {
+    double lower_bound = lower_bounds_.at(row_index);
+    double upper_bound = upper_bounds_.at(row_index);
+
+    if (lower_bound != -kInfinity) {
+      return lower_bound;
+    }
+    return upper_bound;
+  }
+
+  int Presolve::getVariableFeasibleValueMIP(
+    int row_index, int col_index, 
+    int variable_coefficient, double constraint_RHS
+  ) {
+    // Check if RHS/coeff is not an integer
+
+    if (static_cast<int>(constraint_RHS) % variable_coefficient != 0) {
+      // If inequality, check if rounding up satisfies upper
+      // bound and implied bounds.
+      if (row_index < inequalities_count_) {
+        int feasible_value = ceil(constraint_RHS/variable_coefficient);
+        if (feasible_value <= upper_bounds_.at(row_index)) {
+          if (checkVariableImpliedBounds(col_index, feasible_value)) {
+            return feasible_value;
+          }
+        }
+      }
+      return kInfinity;
+    }
+    // If RHS/coeff is integer, calculate it and check if 
+    // implied bounds satisfied.
+    int feasible_value = constraint_RHS/variable_coefficient;
+    if (checkVariableImpliedBounds(col_index, feasible_value)) {
+      return feasible_value;
+    }
+    return kInfinity;
+  };
+
+  bool Presolve::checkIsRowFree(int row_index) {
+    if (lower_bounds_.at(row_index) == -kInfinity && upper_bounds_.at(row_index) == kInfinity) {
+      return true;
+    }
+    return false;
+  }
+
+  void Presolve::updateStateFreeRow(int row_index) {
+    presolve_active_rows_.at(row_index) = false;
+    presolve_active_rows_count_ -= 1;
+
+    // Update presolve stack.
+    struct presolve_log log = {row_index, -1, 0};
+    presolve_stack_.push(log);
+  };
+
+  void Presolve::applyFreeRowPostsolve(int row_index, int col_index) {
+    postsolve_active_rows_.at(row_index) = true;
+  };
+
+  void Presolve::updateStateSingletonVariable(
+    int row_index, int col_index
+  ) {
+    presolve_active_rows_.at(row_index) = false;
+    presolve_active_columns_.at(col_index) = false;
+    presolve_active_rows_count_ -= 1;
+    presolve_active_columns_count -= 1;
+
+    // Update presolve stack.
+    struct presolve_log log = {row_index, col_index, 1};
+    presolve_stack_.push(log);
+  };
+
+  void Presolve::applySingletonVariablePostsolve(
+    int row_index, int col_index
+  ) {
+    int variable_coeff = problem_matrix_.at(row_index).at(col_index);
+    double feasibleValueCalculationBound = getFeasibleValueCalculationBound(
+      row_index
+    );
+    double feasible_value;
+
+    if (solve_MIP_) {
```

Comment:
naming: `solve_mip_` (data variables should be snake case)

Reply:
Isn't this snake case?